README.rst
Codename: ALiSS
Keywords: "Link suggestions agent", "Link agent", Semantic web, RDF, SKOS


ALiSS stands for "Adaptive Links Suggestion Service".
The idea is to automate links relation between webpages via the use of taxonomies and thesauri in combination with the 
popularity of google search results.

The problem:
On a website there is a need to relate pages with eachother depending on the content. This task is normally done 
manually by experts or webmasters. It is an expensive task which requires a carefull attention and it is also 
technically difficult. The "link expert" is supposed to know basic html knowledge. Moreover the manual task of adding 
links leads after some time to a large ammount of links to be manually maintained and updated. This leads at the end to 
a unsustainable situation of many "broken links", website detoriation and finally visitors unsatisfaction. For this 
reason some website try to minimize the use of links, especially external links, reducing this way the basic feature of 
the web: links.

Goal:
The creation and maintenance of links/relations has to be done centrally and 
in a systematic way. If a page does not exist anymore, all the links to this page will be automatically removed by the 
"Link agent". If the page content change drastically than the link to that page should be reconsidered and re-evaluated 
if still of relevance. If we add a new page, with new fresh content, then all the link to older contents should be 
updated to the new fresh page "re-linked". All these tasks are very expensive for medium and large websites, and cannot 
be done manually. The "Link Agent" will take care of these tasks and only leave a small ammount of configuration tasks 
to the web administrator / webmaster.

Technical architecture:
ALiSS is based on the following technologies.

Programming language: python
Presentation / Template language: HTML, DTML, Page templates and CSS
Information language: XML (RDF/SKOS) and OO database objects
Information protocols: HTTP
Web service API: XML-RPC, SOAP
CMS: Zope


System requirements:
TODO: add minimal requirements and dependencies.

Example on how to use it:
ALiSS can be used both as a "taxonomy" server (A-Z index) and with a set of clients (Flash or Ajax)
 which display related links. See howto:s for more info on how to create your client.
TODO: add some real examples.

Contributors:

   Alexandru Ghica - Eau de Web
   Antonio De Marinis - EEA
   Sasha Vinčić - Systemvaruhuset


Epoz/CHANGES.txt
Epoz - Changes

 0.9.0
 
     - Reorganized directory layout into core, i18n, images & plone. Please
       note: You need to "reinstall" Epoz with Quickinstaller / External Method
       when using Epoz in CMF/Plone.
     
	 - Andy Dustman provided a patch for removing MS-Word-Garbage-tags.
	   Thanks Andy!!!
	   
     - Documentation about Epoz for end-users is provided by Tom Purl (see 
       "EpozBook":http://zopewiki.org/EpozBook). Thanks Tom!
 
     - fixed bug: Plone-Toolbox didn't check permissions 'View' and 
       'List folder contents' in breadcrumbs. Thanks to Cyrille Lebeaupin
       from Ingeniweb for patch!
       
     - fixed bug: Image-Preview in Plone-Toolbox was broken in IE.
     
     - fixed bug: JavaScript in content-area could break Epoz...       
       Thanks to Christian Zagrodnick from gocept.com for patch!
       
     - cleaned up line-endings. No more need to write complicated 
       PERL/C++/ADA-Scripts to get rid of them...
     
     
 0.8.6

     - fixed bug: Mozilla / Firefox leaves a single br-tag for
       an empty form. Epoz will clean this mess now.

     - added a wysiwyg_support-switch so Epoz can coexist with 
       other wysiwyg-editors in Plone (esp. Kupu). Moved former
       wysiwyg_support to epoz_wysiwyg_support. Thanks to
       Duncan Booth for patch!!!

     - little cleanup to epoz_script_widget: refactored img-string
       for spacer, so the code becomes more readable.
                
     - fixed bug in epoz_toolbox: CMFPhotos were only shown for preview,
       but not for inserting as image. Thanks to Bjorn van der Meer
       for patch!!!
       
     - updated russian translation; thanks to Michael Krishtopa!
	 
     - added some new FAQs.

 	 
 0.8.5
      
     - added keyboard shortcuts for Mozilla / Firefox which are already
       supported by IE: CTRL + b for bold, CTRL + i for italics, CTRL + u
       for underlined, CTRL + k for create link.

     - added a HowTo for easier setup of mxTidy to get valid XHTML from Epoz
       (see mxTidyHowTo.txt).
     
     - fixed bug: made width/height-settings for Epoz working
       with Archetypes. Thanks to Christian Theune from Gocept for patch!
       
	 - fixed bug: corrected the slashing of folders in Epoz-toolbox 
	   for Plone (http://plone.org/collector/3479)
	          
       
 0.8.4

     - fixed bug: toolbox didn't start from the correct location if the
       portal_factory was activated.

     - fixed bug: wysiwyg support in Plone could be broken, because
       URL3 in wysiwyg_support isn't available under all circumstances.
       I forgot that the damn test()-method evaluates all expressions
       always, regardless of the condition... Thanks to Simon Pamies
       for patch!
       
     - updated browser-detection-javascript to Browser Detect v2.1.6.
              
     - some whitespace-cleanups thanks to AnyEdit for Eclipse.      

 0.8.3

     - added a FAQ. Pleae read it before asking questions...

     - added label for html-source-switch, because many people asked for
       it. Hope this helps clearing up the meaning of the little checkbox
       below the wysiwyg-area...

     - added translations for tidy-messages in the browser-status-bar to
       en & de. Please note: Don't use any entities if you want to translate,
       because the browser-status-bar doesn't support entities!

     - Fixed bug: missing form_path in epoz_script_widget caused problems
       when using portal_factory in Plone. Thanks to Danny Bloemendaal (for
       detecting the problem) and Geoff Davis for patch! Added a check to
       wysiwyg_support to detect document-creation via portal_factory.

     - Fixed bug: the Epoz-Iframe in epoz_script_widget wasn't loaded via
       an absolute url, so browsers could complain when running Epoz via https.

 0.8.2

     - Updated license to Zope Public License 2.1.

     - Danny Bloemendaal provided a cleaned up toolbox for use in Plone 2.
       Thank you, Danny!

     - Added "Photo" (CMFPhoto) as selectable image-type to the toolbox.
       Thanks to Shane Graber for patch!

     - set wordwrap for mxTidy/uTidy to 79 character to avoid some problems
       with splitted a-href-tags.

     - fixed bug: making relative urls for images was broken in EpozPostTidy
       for people who have no mxTidy / uTidylib installed.

     - fixed bug: removed .strip() from data-handler in Extensions/EpozPostTidy.py,
       so required whitespace isn't killed; added a "useless-whitespace"-cleanup
       to EpozPostTidy.

 0.8.1

     - Now the setup for getting relative urls from Epoz is much more easier:
       There's a default EpozPostTidy.py in Epoz/Extensions which can be
       used out of the box.
       Simply create an External-Method in the root of the ZMI or your Plone-Site with
       id=EpozPostTidy, Module-Name=Epoz.EpozPostTidy, Function-Name=EpozPostTidy.

     - added a customcss-parameter (esp. for use in Plone) to render customized
       css-styles in difference to a global css correctly.

     - fixed bug: the XMLRPC for tidying etc. is now independent of your
       Zope-Server-Root, so the EpozPostTidy-Hook can be created in
       your Plone-Instance without a VHM-Setup anymore.

     - fixed bug: canceling "Add anchor" resulted in emtpy tag. Thanks to
       Simon Pamies for patch!!!

     - fixed bug: IE complained about security when Epoz was called via
       https. Thanks to Sam Brauer and Andrew Athan for patches!

     - fixed bug: <embed>-tag was replaced by <ibed> in Mozilla. Thanks
       to Christian Ledermann.

     - fixed bug in epoz_script_main.js.dtml/InitEpoz: IE 5.0 isn't
       wysiwyg-able and therefore needs to fall back to a textarea. Thanks
       to Marc Zacher!

     - updated epoz_lang_pt-br. Thanks to Luiz Fernando Pasqual S. Souza!

     - updated epoz_lang_zh-cn. Thanks to panjunyong!

     - updated epoz_lang_no.js.dtml. Thanks to Einar Jensen!


 0.8.0


Epoz/__init__.py
#####
###  Epoz - a cross-browser-wysiwyg-editor for Zope
##   Copyright (C) 2005 Maik Jablonski (maik.jablonski@uni-bielefeld.de)
#
#    All Rights Reserved.
#
#    This software is subject to the provisions of the Zope Public License,
#    Version 2.1 (ZPL). A copy of the ZPL should accompany this distribution.
#    THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
#    WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
##   WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
###  FOR A PARTICULAR PURPOSE.
####

import re
import urlparse
import cgi

# try to import mxTidy as default
#try:
#    from mx import Tidy
#    mxTidyIsAvailable = 1
#except:
mxTidyIsAvailable = 0

# try to import uTidylib as fallback
try:
    import tidy
    uTidyIsAvailable = 1
except ImportError:
    uTidyIsAvailable = 0

import Globals
from Products.PageTemplates.PageTemplateFile import PageTemplateFile
from ImageFile import ImageFile

# Regexp for snipping the body of a document
HTML_BODY = re.compile('<body[^>]*?>(.*)</body[^>]*?>', re.DOTALL|re.IGNORECASE)
MSO_CLASS = re.compile(r'<(\w+)\W+(class="Mso.*?")>', re.IGNORECASE)
EPOZ_SCRIPT = re.compile(r'<epoz:script (style="[^"]*")? ?')

misc_ = {
 # The Format-Controls
'epoz_script_main.js':ImageFile('epoz/epoz_core/epoz_script_main.js.dtml', globals()),

 # Rendering of the widget
'epoz_script_widget.js':ImageFile('epoz/epoz_core/epoz_script_widget.js.dtml', globals()),

 # Browser-detection
'epoz_script_detect.js':ImageFile('epoz/epoz_core/epoz_script_detect.js.dtml', globals()),

 # Code for the color-selection-popup
'epoz_script_color.html':ImageFile('epoz/epoz_core/epoz_script_color.html.dtml', globals()),

 # Code for the table-popup
'epoz_script_table.html':ImageFile('epoz/epoz_core/epoz_script_table.html.dtml', globals()),

 # The XMLRPC-handler
'vcXMLRPC.js':ImageFile('epoz/epoz_core/vcXMLRPC.js.dtml', globals()),

 # The Epoz-Language-Files
'epoz_lang_da.js':ImageFile('epoz/epoz_i18n/epoz_lang_da.js.dtml', globals()),
'epoz_lang_de.js':ImageFile('epoz/epoz_i18n/epoz_lang_de.js.dtml', globals()),
'epoz_lang_en.js':ImageFile('epoz/epoz_i18n/epoz_lang_en.js.dtml', globals()),
'epoz_lang_es.js':ImageFile('epoz/epoz_i18n/epoz_lang_es.js.dtml', globals()),
'epoz_lang_fi.js':ImageFile('epoz/epoz_i18n/epoz_lang_fi.js.dtml', globals()),
'epoz_lang_fr.js':ImageFile('epoz/epoz_i18n/epoz_lang_fr.js.dtml', globals()),
'epoz_lang_hu.js':ImageFile('epoz/epoz_i18n/epoz_lang_hu.js.dtml', globals()),
'epoz_lang_it.js':ImageFile('epoz/epoz_i18n/epoz_lang_it.js.dtml', globals()),
'epoz_lang_nl.js':ImageFile('epoz/epoz_i18n/epoz_lang_nl.js.dtml', globals()),
'epoz_lang_no.js':ImageFile('epoz/epoz_i18n/epoz_lang_no.js.dtml', globals()),
'epoz_lang_pl.js':ImageFile('epoz/epoz_i18n/epoz_lang_pl.js.dtml', globals()),
'epoz_lang_pt.js':ImageFile('epoz/epoz_i18n/epoz_lang_pt.js.dtml', globals()),
'epoz_lang_pt-br.js':ImageFile('epoz/epoz_i18n/epoz_lang_pt-br.js.dtml', globals()),
'epoz_lang_ru.js':ImageFile('epoz/epoz_i18n/epoz_lang_ru.js.dtml', globals()),
'epoz_lang_zh-cn.js':ImageFile('epoz/epoz_i18n/epoz_lang_zh-cn.js.dtml', globals()),

 # The Epoz-Buttons
'epoz_button_anchor.gif':ImageFile('epoz/epoz_images/epoz_button_anchor.gif', globals()),
'epoz_button_bgcolor.gif':ImageFile('epoz/epoz_images/epoz_button_bgcolor.gif', globals()),
'epoz_button_bold.gif':ImageFile('epoz/epoz_images/epoz_button_bold.gif', globals()),
'epoz_button_centre.gif':ImageFile('epoz/epoz_images/epoz_button_centre.gif', globals()),
'epoz_button_hr.gif':ImageFile('epoz/epoz_images/epoz_button_hr.gif', globals()),
'epoz_button_hyperlink.gif':ImageFile('epoz/epoz_images/epoz_button_hyperlink.gif', globals()),
'epoz_button_image.gif':ImageFile('epoz/epoz_images/epoz_button_image.gif', globals()),
'epoz_button_indent.gif':ImageFile('epoz/epoz_images/epoz_button_indent.gif', globals()),
'epoz_button_italic.gif':ImageFile('epoz/epoz_images/epoz_button_italic.gif', globals()),
'epoz_button_left_just.gif':ImageFile('epoz/epoz_images/epoz_button_left_just.gif', globals()),
'epoz_button_list.gif':ImageFile('epoz/epoz_images/epoz_button_list.gif', globals()),
'epoz_button_numbered_list.gif':ImageFile('epoz/epoz_images/epoz_button_numbered_list.gif', globals()),
'epoz_button_outdent.gif':ImageFile('epoz/epoz_images/epoz_button_outdent.gif', globals()),
'epoz_button_redo.gif':ImageFile('epoz/epoz_images/epoz_button_redo.gif', globals()),
'epoz_button_right_just.gif':ImageFile('epoz/epoz_images/epoz_button_right_just.gif', globals()),
'epoz_button_strikethrough.gif':ImageFile('epoz/epoz_images/epoz_button_strikethrough.gif', globals()),
'epoz_button_subscript.gif':ImageFile('epoz/epoz_images/epoz_button_subscript.gif', globals()),
'epoz_button_superscript.gif':ImageFile('epoz/epoz_images/epoz_button_superscript.gif', globals()),
'epoz_button_table.gif':ImageFile('epoz/epoz_images/epoz_button_table.gif', globals()),
'epoz_button_textcolor.gif':ImageFile('epoz/epoz_images/epoz_button_textcolor.gif', globals()),
'epoz_button_tools.gif':ImageFile('epoz/epoz_images/epoz_button_tools.gif', globals()),
'epoz_button_underline.gif':ImageFile('epoz/epoz_images/epoz_button_underline.gif', globals()),
'epoz_button_undo.gif':ImageFile('epoz/epoz_images/epoz_button_undo.gif', globals()),
'epoz_button_unformat.gif':ImageFile('epoz/epoz_images/epoz_button_unformat.gif', globals()),
}


def Epoz(self, name, data='', toolbox='', lang='en',
               path='', widget='',
               style='width: 600px; height: 250px; border: 1px solid #000000;',
               button='background-color: #EFEFEF; border: 1px solid #A0A0A0; cursor: pointer; margin-right: 1px; margin-bottom: 1px;',
               css='', customcss='', charset='utf-8', pageurl=''):
    """ Create an Epoz-Wysiwyg-Editor.

        name : the name of the form-element which submits the data
        data : the data to edit
        toolbox : a link to a HTML-Page which delivers additional tools
        lang: a code for the language-file (en,de,...)
        path : path to Epoz-Javascript. Needed mainly for Plone (portal_url).
        widget: You can specify a path to an alternative JavaScript for
                epoz_script_widget.js
        style : style-definition for the editor-area
        button : style-definiton for buttons
        css : url to a global css which should be used for rendering
               content in editor-area
        customcss: url to a customized css which should be used for rendering
                        content in editor-area
        charset : charset which is edited in the editor-area
        pageurl: the base-url for the edited page

        If Epoz can't create a Rich-Text-Editor, a simple textarea is created.
    """

    if data is None:
        data=''

    js_data = data
    data = cgi.escape(data)

    # hide scripts without removing them
    js_data = js_data.replace('<script ',
                                    '<epoz:script style="display: none" ')
    js_data = js_data.replace('</script>', '</epoz:script>')

    # Quote newlines and single quotes, so the Epoz-JavaScript won't break.
    # Needs to be a list and no dictionary, cause we need order!!!

    quotes = (("\\","\\\\"), ("\n","\\n"), ("\r","\\r"), ("'","\\'"))

    for item in quotes:
        js_data = js_data.replace(item[0], item[1])

    # Determine the correct path for VirtualHostMonsters & PCGI & etc.pp.
    if not path:
        path = "%s/misc_/Epoz/" % (self.REQUEST.get('BASEPATH1',''),)

    if not pageurl:
        pageurl = self.REQUEST.get('URL1','')

    i18n='<script language="JavaScript" type="text/javascript" src="%sepoz_lang_en.js"></script>' % (path,)
    if lang<>'en':
        i18n += '<script language="JavaScript" type="text/javascript" src="%sepoz_lang_%s.js"></script>' % (path, lang)

    if not widget:
       widget = "%sepoz_script_widget.js" % path

    # This is just a dummy page for test-iframe
    # to prevent IE complaining when using Epoz via https
    iframesrc = pageurl+'/epoz_blank_iframe.html'

    # Return the HTML-Code for the Epoz-Rich-Text-Editor
    return """
<iframe id="EpozIFrame" src="%(iframesrc)s" style="position: absolute; visibility: hidden; width: 0px; height: 0px;"></iframe>
%(i18n)s
<script language="JavaScript" type="text/javascript" src="%(widget)s"></script>
<script language="JavaScript" type="text/javascript" src="%(path)svcXMLRPC.js"></script>
<script language="JavaScript" type="text/javascript" src="%(path)sepoz_script_detect.js"></script>
<script language="JavaScript" type="text/javascript" src="%(path)sepoz_script_main.js"></script>
<script language="JavaScript" type="text/javascript">
<!--
InitEpoz('%(name)s','%(js_data)s','%(path)s','%(toolbox)s','%(style)s','%(button)s','%(css)s','%(customcss)s','%(charset)s', '%(pageurl)s');
//-->
</script>
<noscript><textarea name="%(name)s" style="%(style)s">%(data)s</textarea></noscript>
""" % {
       'i18n': i18n,
       'widget':  widget,
       'path':    path,
       'name':    name,
       'js_data': js_data,
       'toolbox': toolbox,
       'style':   style,
       'button':  button,
       'css': css,
       'customcss': customcss,
       'charset': charset,
       'data':    data,
       'pageurl': pageurl,
       'iframesrc': iframesrc
      }


def EpozTidy(self, html, pageurl):
    """ Take html and deliver xhtml if mxTidy is installed;
        call EpozPostTidy for html-postprocessings before returning the result
    """

    errors = 0
    output = html
    errordata = ""

    input = html.encode("utf-8")
    input = EPOZ_SCRIPT.sub('<script ', input)
    input = input.replace('</epoz:script>', '</script>')

    if mxTidyIsAvailable:
        (errors, warnings, output, errordata) = Tidy.tidy(
            input, drop_empty_paras=1, indent_spaces=1, indent="auto",clean=1,
            output_xhtml=1, word_2000=1, drop_font_tags=1, wrap=79, char_encoding="utf8")
        if errors:
            output = html
    elif uTidyIsAvailable:
        parsed = tidy.parseString(
            input, drop_empty_paras=1, indent_spaces=1, indent="auto",
            output_xhtml=1, word_2000=1, wrap=79, char_encoding="utf8",
            add_xml_decl=0, doctype="omit", indent_attributes=1,
            drop_proprietary_attributes=1, bare=1, clean=1,
            enclose_text=1, tidy_mark=0)
        reports = parsed.get_errors()
        all_errors = [str(x) for x in reports if x.severity != 'W']
        errors = len(all_errors)
        errordata = '\n'.join(all_errors)
        if errors:
            output = html
        else:
            output = str(parsed)

    output = MSO_CLASS.sub(r"<\1>", output)
    result = HTML_BODY.search(output)
    if result:
        output = result.group(1)

    # Call External Method / PythonScript for postprocessing
    # The script should expect two parameters:
    # self = called context (=server)
    # html = the htmlbody to postprocess
    # pathname = path of edited object (maybe with template!)
    # The script should return the new htmlbody

    EpozPostTidy = getattr(self, 'EpozPostTidy', None)
    if EpozPostTidy is not None:
        output = EpozPostTidy(self, output, pageurl)

    return (errors, output, errordata)


def EpozGetRelativeUrl(self, pageUrl, targetUrl):
    """ Returns relative path from targetUrl to pageUrl. """

    # Just some shortcuts...
    SCHEME, HOSTNAME, PATH, PARAMS, QUERY, FRAGMENT = range(6)

    pageUnits = urlparse.urlparse(pageUrl)
    targetUnits = urlparse.urlparse(targetUrl)

    if pageUnits[:PATH] != targetUnits[:PATH]:
        # Scheme (http) or host:port (www.comain.com) are different
        # -> no possible relative URL
        return targetUrl

    afterPath = targetUnits[PARAMS:]

    pagePath = pageUnits[PATH].split('/')
    pagePath.reverse()

    targetPath = targetUnits[PATH].split('/')
    targetPath.reverse()

    # Remove parts which are equal (['a', 'b'] ['a', 'c'] --> ['c'])
    while (len(pagePath) > 0
       and len(targetPath) > 0
       and pagePath[-1] == targetPath[-1]):
        pagePath.pop()
        last_pop = targetPath.pop()
    if len(pagePath) == 0 and len(targetPath) == 0:
        # Special case: link to itself (http://foo/a http://foo/a -> a)
        targetPath.append(last_pop)
    if len(pagePath) == 1 and pagePath[0] == "" and len(targetPath) == 0:
        # Special case: (http://foo/a/ http://foo/a -> ../a)
        pagePath.append(last_pop)
        targetPath.append(last_pop)

    targetPath.reverse()

    relativeUrl = ['..'] * (len(pagePath) -1)
    relativeUrl = ('', '', '/'.join(relativeUrl + targetPath)) + tuple(afterPath)

    return urlparse.urlunparse(relativeUrl)


# Make sure the Epoz method is replaceable in the Products folder
Epoz.__replaceable__ = Globals.REPLACEABLE

# Register Epoz as global method
# Register EpozTidy as global method for XMLRPC
# Register EpozGetRelativeUrl as global method
# Register epoz_blank_iframe.html as global PageTemplate

methods = {
'Epoz': Epoz,
'EpozTidy': EpozTidy,
'EpozGetRelativeUrl': EpozGetRelativeUrl,
'epoz_blank_iframe.html': PageTemplateFile('epoz/epoz_core/epoz_blank_iframe.html.pt',
                          globals()) ,
}


# And now try to register Epoz for CMF/Plone

try:
    from Products.CMFCore.DirectoryView import registerDirectory

    global cmfepoz_globals
    cmfepoz_globals=globals()

    def initialize(context):
        registerDirectory('epoz', cmfepoz_globals)
except:
    pass
    
